### PR TITLE
JIT: workaround for unreliable change detection in fgReorderBlocks

### LINKED ISSUE #48495: Assertion failed 'm_compGenTreeID == m_compiler->compGenTreeID'
After https://github.com/dotnet/runtime/pull/48364, spmi replay of libraries.crossgen2.windows.x64.checked.mch fails with:
```
ERROR: Exception thrown: DebugBreak or AV Exception 123
ERROR: main method 28 of size 134 failed to load and compile correctly.
ERROR: Exception thrown: DebugBreak or AV Exception 123
ERROR: main method 24 of size 134 failed to load and compile correctly.
ISSUE: <ASSERT> D:\workspace\_work\1\s\src\coreclr\jit\phase.cpp (47) - Assertion failed 'm_compGenTreeID == m_compiler->compGenTreeID' in '<System-Collections-Generic-IEnumerable<Microsoft-Diagnostics-Tracing-Analysis-TraceProcess>-GetEnumerator>d__8:MoveNext():bool:this' during 'Optimize layout' (IL size 134)
ISSUE: <ASSERT> D:\workspace\_work\1\s\src\coreclr\jit\phase.cpp (47) - Assertion failed 'm_compGenTreeID == m_compiler->compGenTreeID' in '<System-Collections-IEnumerable-GetEnumerator>d__26:MoveNext():bool:this' during 'Optimize layout' (IL size 134)
```

Also, due to the same PR, libraries.crossgen.windows.x64.checked.mch has a single new "missing" method (possibly expected?)
```
MISSING: Method context 32558 failed to replay: SuperPMI assertion 'IsFieldStatic->GetIndex(CastHandle(fhld)) != -1' failed ("Didn't find 00000298B1E294A8")
```

@AndyAyersMS PTAL
cc @dotnet/jit-contrib

### PR DESCRIPTION
**Pull request by AndyAyersMS**
This method costs trees, which can in turn modify the IR by swapping operands.
As a result the bool value returned doesn't properly reflect whether any
changes happened.

This impacts proper reporting phase status by `optOptimizeLayout'. Since phase
status just gates post-phase dumps and checks, we'll just claim this phase
always modifies IR.

Fixes #48495.

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
cc @dotnet/jit-contrib

**Review by BruceForstall — APPROVED** [review]

**Comment by AndyAyersMS** [comment]
Looks like the preceding phase (invert loops) may have the same problem, given #48494.

So will look to extend this fix to cover that case too.

**Comment by AndyAyersMS** [comment]
Kind of on the fence here. On the one had it would be nice to be able to cost trees without side effecting them; on the other hand an accurate costing (at least as accurate as it is ever going to be) more or less entails opportunistic swapping (and address mode marking).

So three options to pursue... I have chosen the simplest for now:
* assume these phases always change IR (the only downside is extra dumps / checks in non-release)
* update costing APIs to allow a side-effect free mode
* update costing APIs to return whether they caused any side effects (other than setting costs)

**Comment by AndyAyersMS** [comment]
I suppose there is a fourth option:
* use the logic the current post-phase debug check as a way of detecting that a phase made changes. In particular the check that is firing here is that some new tree node was created.

I'm more in favor of having phases be deliberate in what they change, so decided not to pursue this one.

None of this is crucial for now, but at some point we may want to report finer-grained phase status and relying on to guide release behaviors.

**Comment by BruceForstall** [comment]
I somewhat prefer the 3rd:
> update costing APIs to return whether they caused any side effects (other than setting costs)
as long as it can also be checked somehow. If it's expensive, though, then going "conservative" and assuming changes doesn't seem too bad an alternative.

File: src/coreclr/jit/optimizer.cpp [hunk 1]
```diff
@@ -4493,6 +4493,15 @@ PhaseStatus Compiler::optInvertLoops()
...
 
+    // optInvertWhileLoop can cause IR changes even if it does not modify
+    // the flow graph. It calls gtPrepareCost which can cause operand swapping.
+    // Work around this for now.
+    //
+    // Note phase status only impacts dumping and checking done post-phase,
+    // it has no impact on a release build.
+    //
+    madeChanges = true;
+
```
Comment (briansull) [inline src/coreclr/jit/optimizer.cpp:4504]: You could also add an additional state MODIFIED_TREE here instead of using MODIFIED_EVERYTHING

**Review by briansull — APPROVED** [review]
Looks OK

File: src/coreclr/jit/optimizer.cpp (hunk 1 above)
Comment (AndyAyersMS) [reply src/coreclr/jit/optimizer.cpp:4504]: Right -- I originally had finer-grained status in mind, but wasn't quite sure what the right breakout should be, so just have everything/nothing for now.

**Comment by trylek** [comment]
Great, thank you! It would be great if you could run this change through the shiproom and push it to Preview 2 that's just forked.

**Comment by AndyAyersMS** [comment]
> It would be great if you could run this change through the shiproom 

Sure.

**Comment by AndyAyersMS** [comment]
I'm fairly sure the outerloop failures are unrelated to this change. But I'm reluctant to merge with things looking this red.

Lots of instances of `tracing\\eventpipe\\eventsvalidation\\ThreadPool\\ThreadPool` failing, likely #48543.

And some pal test failures on Linux, dups of #48496.

And a jit-format failure on Linux, looks like jit-format is crashing; the patch artifact is a zero sized file.

**Comment by AndyAyersMS** [comment]
Local run of jit-format on linux was clean.

**Comment by AndyAyersMS** [comment]
Looked again at the failures and don't see anything related. So am going to merge after all.

**Pull request merged**